rust/lance-encoding/src/encodings/physical/rle.rs

```diff
@@ -1016,4 +955,79 @@ mod tests {
             }
         }
     }
+
+    // ========== Encoding Verification Tests ==========
+
+    #[test_log::test(tokio::test)]
+    async fn test_rle_encoding_verification() {
+        use crate::testing::{check_round_trip_encoding_of_data, TestCases};
+        use crate::version::LanceFileVersion;
+        use arrow_array::{Array, Int32Array};
+        use lance_datagen::{ArrayGenerator, RowCount};
+        use std::collections::HashMap;
+        use std::sync::Arc;
+
+        let test_cases = TestCases::default()
+            .with_expected_encoding("rle")
+            .with_file_version(LanceFileVersion::V2_1);
+
+        // Test both explicit metadata and automatic selection
+        // 1. Test with explicit RLE threshold metadata
+        let metadata_explicit = HashMap::from([(
+            "lance-encoding:rle-threshold".to_string(),
+            "0.8".to_string(),
+        )]);
+        let mut generator = RleDataGenerator::new(vec![1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3]);
+        let data_explicit = generator.generate_default(RowCount::from(10000)).unwrap();
+        check_round_trip_encoding_of_data(vec![data_explicit], &test_cases, metadata_explicit)
+            .await;
+
+        // 2. Test automatic RLE selection based on data characteristics
+        // 80% repetition should trigger RLE (> default 50% threshold)
+        let mut values = vec![42i32; 8000]; // 80% repetition
+        values.extend([1i32, 2i32, 3i32, 4i32, 5i32].repeat(400)); // 20% variety
+        let arr = Arc::new(Int32Array::from(values)) as Arc<dyn Array>;
+        check_round_trip_encoding_of_data(vec![arr], &test_cases, HashMap::new()).await;
+    }
+
+    /// Generator that produces repetitive patterns suitable for RLE
+    #[derive(Debug)]
+    struct RleDataGenerator {
+        pattern: Vec<i32>,
+        idx: usize,
+    }
+
+    impl RleDataGenerator {
+        fn new(pattern: Vec<i32>) -> Self {
+            Self { pattern, idx: 0 }
+        }
+    }
+
+    impl lance_datagen::ArrayGenerator for RleDataGenerator {
+        fn generate(
+            &mut self,
+            _length: lance_datagen::RowCount,
+            _rng: &mut rand_xoshiro::Xoshiro256PlusPlus,
+        ) -> std::result::Result<std::sync::Arc<dyn arrow_array::Array>, arrow_schema::ArrowError>
+        {
+            use arrow_array::Int32Array;
+            use std::sync::Arc;
+
+            // Generate enough repetitive data to trigger RLE
+            let mut values = Vec::new();
+            for _ in 0..10000 {
+                values.push(self.pattern[self.idx]);
+                self.idx = (self.idx + 1) % self.pattern.len();
+            }
+            Ok(Arc::new(Int32Array::from(values)))
+        }
+
+        fn data_type(&self) -> &arrow_schema::DataType {
+            &arrow_schema::DataType::Int32
+        }
+
+        fn element_size_bytes(&self) -> Option<lance_datagen::ByteCount> {
+            Some(lance_datagen::ByteCount::from(4))
+        }
+    }
```

Comment:
Can we keep this in the `lance-datagen` package?  It seems generally useful.  Also, we might be able to implement it as a generic "repeat" generator that repeats elements from an underlying generator (either in fixed run lengths or random run lengths)

Can do in a follow-up too

Reply:
Tracked in https://github.com/lancedb/lance/issues/4392